Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,254 @@
+            return allFiles;
+        }
+    }
+
+    tl.debug('using: ' + filePatterns.length + ' filePatterns: ' + filePatterns + ' to search for files.');
+
+    // minimatch options
+    var matchOptions = { matchBase: true };
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug('Candidates found for match: ' + allFiles.length);
+    for (var i = 0; i < allFiles.length; i++) {
+        tl.debug('file: ' + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    var SortedSet = require('collections/sorted-set');
+    var matchingFilesSet = new SortedSet();
+
+    for (var i = 0; i < filePatterns.length; i++) {
+        var normalizedPattern: string = path.join(rootFolder, path.normalize(filePatterns[i]));
+
+        tl.debug('searching for files, pattern: ' + normalizedPattern);
+
+        var matched = tl.match(allFiles, normalizedPattern, matchOptions);
+        tl.debug('Found total matches: ' + matched.length);
+        // ensure each result is only added once
+        for (var j = 0; j < matched.length; j++) {
+            var match = path.normalize(matched[j]);
+            if (matchingFilesSet.add(match)) {
+                var stats = tl.stats(match);
+                tl.debug('adding ' + (stats.isFile() ? 'file:   ' : 'folder: ') + match);
+                if (stats.isFile()) {
+                    var parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug('adding folder:' + parent);
+                    }
+                }
+            }
+        }
+    }
+    return matchingFilesSet.sorted();
+}
+
+var remotePath = tl.getInput('remotePath', true).trim();
+
+var Client = require('ftp');
+var c = new Client();
+
+var files = findFiles();
+
+var filesUploaded: number = 0;
+var filesSkipped: number = 0; // already exists and overwrite mode off
+var directoriesCreated: number = 0;
+var directoriesSkipped: number = 0; // already exists
+
+function checkDone(): void {
+    var total: number = filesUploaded + filesSkipped + directoriesCreated + directoriesSkipped;
+    var remaining: number = files.length - total;
+    tl.debug(
+        'filesUploaded: ' + filesUploaded +
+        ', filesSkipped: ' + filesUploaded +
+        ', directoriesCreated: ' + directoriesCreated +
+        ', directoriesSkipped: ' + directoriesSkipped +
+        ', total: ' + total + ', remaining: ' + remaining);
+    if (remaining == 0) {
+        c.end();
+        tl.setResult(tl.TaskResult.Succeeded,
+            'Ftp upload successful' +
+            '\nhost: ' + serverEndpointUrl.host +
+            '\npath: ' + remotePath +
+            '\n files uploaded: ' + filesUploaded +
+            '\n files skipped: ' + filesSkipped +
+            '\n directories created: ' + directoriesCreated +
+            '\n directories skipped: ' + directoriesSkipped
+        );
+    }
+}
+
+
+function uploadFiles() {
+    var Set = require('collections/set');
```

Comment:
Can this require('collections/set') be an import statement at the top of the file?